create function of the DB

josephShokry opened the pull request.

josephShokry committed:
parsing input using cmd

adding example of how to parse input using cmd

josephShokry committed:
handling json

adding example of handling json data

josephShokry committed:
adding the main file

adding the main file parsing the input (the command and the json) and check the command then execute the proper function

josephShokry committed:
adding more input variables

josephShokry committed:
adding the create function

- adding the create function that makes the directories using the schema
- modifying how to get the schema instead of getting it from the cmd directly now we get it through a text file

josephShokry committed:
create funciton

completing the create function, making the file of the parser and cleaning the code

josephShokry committed:
handling errors of invalid schema file

if the user parsed a not existent file of not parsing the schema file at all

josephShokry committed:
Delete __pycache__ directory

josephShokry committed:
adding command factory and help in parser

i merged the code of hussain about the command factory with the parser the main code and adder the help attribute of the parsers

josephShokry committed:
.

josephShokry committed:
Merge branch 'tests-template' of https://github.com/josephShokry/SimpleFSDB into tests-template

josephShokry committed:
adding the keys class

adding the keys class to simplify the coding

josephShokry committed:
adding test cases

adding test cases of create function to check if all the tables and DB has been created or not

hussainmansour committed:
merging between test-template and AvoidingError

josephShokry committed:
Delete __pycache__ directory

josephShokry committed:
Delete tests directory

josephShokry committed:
handling some comments

Commit:
Handling if tables are not found in schema

EhabYasser25 committed:
Delete __pycache__ directory

Commit:
handling schema input errors

handling if schema file is not existed or if input path for schema is wrong or if schema is written in a wrong json format

Commit:
Merge branch 'Handling-Errors' of https://github.com/josephShokry/SimpleFSDB into Handling-Errors

Commit:
command functions folder

Commit:
Making a folder for all commands

Commit:
Making a folder for all commands

Commit:
Handling input wrong format of schema

Commit:
delete extra files

Commit:
Handling some comments

Commit:
Handling some comments

Commit:
minor changes

Commit:
implement validate

Commit:
removing extra files

hussainmansour committed:
make it more functionality

josephShokry committed:
adding the github work flow and the status

josephShokry committed:
testing the work flow and adding more test cases

josephShokry committed:
testing the workflow#3

josephShokry committed:
testing the workflow#4

josephShokry committed:
test github workflow#5

josephShokry committed:
testing github workflow#6

josephShokry committed:
adding the exceptions and cleaning the folders

josephShokry committed:
Create output.py

this file will take any exception and output the right message, now the program will not crash for any cause after modifying the other files as this concept

josephShokry committed:
catch error in main & exception factory

i added the try catch on the main that will catch any exception raised during the execution of the program, after that it will send this exception to the exception factory that will return exception object form it we will get the status and the message and send it to output function.

josephShokry committed:
more test cases and more raise exception

i made one raise exception for the user if not giving the schema path in create command and solved the errors of it.
add one more test case for this condition.
there is still one problem needed to be solved that the attributes of the exception don't get updated with new parameters when calling it with new parameters

hussainmansour committed:
Merge branch 'Handling-Errors' of https://github.com/josephShokry/SimpleFSDB into Handling-Errors

hussainmansour committed:
Merge branch 'Handling-Errors' of https://github.com/josephShokry/SimpleFSDB into Handling-Errors

hussainmansour committed:
some changes in the function

josephShokry committed:
adding raise exceptions ,test cases

we handled the possible errors that may make the program fail
the invalid schema file
the missing of schema file
the wrong command
the wrong schema file
and deleted the exception factory

mohamedabdelsalam commented on Commands_Functions/Command_Factory.py:
```diff
@@ -0,0 +1,21 @@
+from Set_Command import SetCommand
+from output.exceptions import *
+from output.status import Status
+
```
file name should be small letters

mohamedabdelsalam commented on Commands_Functions/Create_Command.py:
```diff
@@ -0,0 +1,33 @@
+class CreateCommand(Icommand):
+    def __init__(self, schema_path):
+        self.schema_path = schema_path
+    def isvalid(self):
```
I recommend make this as a static and return the schema object 
```
__init__(self, schema_path):
    self.schema = CreateCommand.__validate_and_get_schema()
```

mohamedabdelsalam commented on Commands_Functions/Create_Command.py:
```diff
@@ -0,0 +1,33 @@
+        if self.schema_path == None or not os.path.exists(self.schema_path) :
+            raise MissingInput(status = Status.MissingInput, message = "the schema is missing")
+        self.schema_file = self.schema_path
+        schema = open(self.schema_path, "r")
```
use `with open `

mohamedabdelsalam commented on Commands_Functions/Create_Command.py:
```diff
@@ -0,0 +1,33 @@
+        except:
+            raise WrongInput(status = Status.WrongInput, message = "schema is written in a wrong json format") 
+        schema.close()
+        return Status.SUCCESS
```
It shouldn't return the status. If it doesn't then throw an exception, it should be success. 

if it return status, what will GetCommand().execute return ?

mohamedabdelsalam commented on Commands_Functions/Create_Command.py:
```diff
@@ -0,0 +1,33 @@
+        schema.close()
+        return Status.SUCCESS
+
+    def ExcuteInternal(self):
```
method name is small letter

mohamedabdelsalam commented on Commands_Functions/Icommand.py:
```diff
@@ -0,0 +1,18 @@
+from abc import ABC, abstractmethod
```
This should be Abstract command not interface anymore

hussainmansour replied:
we changed it to interface as there is no function implemented on this class

mohamedabdelsalam commented on Commands_Functions/Icommand.py:
```diff
@@ -0,0 +1,18 @@
+"""the interface Icommand"""
+class Icommand(ABC):
+    def execute(self,status):
+        if status == Status.SUCCESS:
+           self.ExcuteInternal()
```
how will we know the status before execute ??!!

mohamedabdelsalam commented on Commands_Functions/Icommand.py:
```diff
@@ -0,0 +1,18 @@
+    def isvalid(self):
+        pass
+    @abstractmethod 
+    def ExcuteInternal(self):
```
execute_internal

mohamedabdelsalam commented on Commands_Functions/Icommand.py:
```diff
@@ -0,0 +1,18 @@
+            return #an error message
+    
+    @abstractmethod                                          
+    def isvalid(self):
```
remove this if you don't call it in AbstractCommand

mohamedabdelsalam commented on output/output.py:
```diff
@@ -0,0 +1,11 @@
+from output.status import Status
+
+
+class OutPut:
```
You shouldn't print anything here.  I already explained the concept of API

hussainmansour replied:
we changed it to print the final message in the main Is this what you mean? if not could you create a meeting with us to understand what you mean? and thanks a lot

mohamedabdelsalam commented on Commands_Functions/Set_Command.py:
```diff
@@ -0,0 +1,12 @@
+    def ExcuteInternal(self):
+        pass
+ 
+ 
```
remove all .pyc files

Commit:
some changes

Commit:
Update main.py

Commit:
check

Commit:
some changes

Commit:
changes

mohamedabdelsalam commented on Commands_Functions/create_command.py:
```diff
@@ -0,0 +1,33 @@
+        self.schema_path = schema_path
+        self.isvalid()
+        
+    def isvalid(self):
```
why this method isn't a static ? 

I recommend to be a static and return the data. It gives me an impression that it is a boolean and it doesn't return boolean.

I recommend 
```
self.schema = CreateCommand.__Validate_and_get_Schema(schema_path)
```

hussainmansour replied:
isvalid(): it is a function that checks all the errors possible and if there is not an error occurs it get the data and pass it to the self parameter and every class command has his own is valid function and must be implemented so it must be in the interface icommand

mohamedabdelsalam commented on Commands_Functions/create_command.py:
```diff
@@ -0,0 +1,33 @@
+                raise WrongInput(status = Status.WrongInput, message = "schema is written in a wrong json format")
+
+    def execute(self):
+        schema_data = self.schema_data
```
why u have `schema_data = self.schema_data`

josephShokry replied:
when I wrote it, I wrote it because I wanted to bring the schema_data variable from the ```init``` to the ```execute``` by assigning it to the variable ```schema_data``` but now after learning more about python class, it is useless as I can say in the ```execute``` function ```self.schema_data```. I solved it.

mohamedabdelsalam commented on Commands_Functions/icommand.py:
```diff
@@ -0,0 +1,13 @@
+    def execute(self):
+        pass
+    
+    @abstractmethod                                          
+    def isvalid(self):
+        pass
```
We don't need this anymore

hussainmansour replied:
please see the previous comment & if there is any misunderstanding could you please create a meeting with us

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,18 @@
+    try:
+        args = parser.parseInput()
+        command = CommandFactory.create_commands(args)
+        command.execute()
```
`result = command.execute()`

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,18 @@
+    except Exception as e:
+        output_object = OutPut(status = e.status, message = e.message)
+    else:
+        output_object = OutPut(status = Status.SUCCESS)
```
pass the result here

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,18 @@
+    else:
+        output_object = OutPut(status = Status.SUCCESS)
+
+print(output_object.message)
```
You should convert output to json and print the output.

josephShokry replied:
done

mohamedabdelsalam commented on output/output.py:
```diff
@@ -0,0 +1,8 @@
+from output.status import Status
+
+class OutPut:
```
Output not OutPut

josephShokry replied:
ok done.

mohamedabdelsalam commented on output/output.py:
```diff
@@ -0,0 +1,8 @@
+from output.status import Status
+
+class OutPut:
+    def __init__(self, result = None, message = None, status = None):
```
You aren't set the result and status in the instance variable, so why are you passing them ??!!! you need to have 

pass an exception here. 
message = str(exception) if excetion not one else "success"
self.status = success if not exception else exception.status

josephShokry replied:
ok, I handled this comment.

mohamedabdelsalam commented on output/exceptions.py:
```diff
@@ -0,0 +1,31 @@
+class MissingInput(Exception):
+    def __init__(self, status = Status.MissingInput, message = "there is a missed required input"):
+        self.status = status
+        self.message = message
```
You don't need self.message here. The parent class has a message as an instance variable.

josephShokry replied:
I don't know the importance of sending the message to the superclass, we prefer to have the exception attributes in the class of it.
so we removed the line of  ```super().__init__(self.message)```

mohamedabdelsalam replied:
You need to read about super in Java. Read about inheritance and why we use super in constructor

hussainmansour committed:
solving comments

changing the output function that will take an excetption and return json object contain status, message and result

mohamedabdelsalam commented on Commands_Functions/create_command.py:
```diff
@@ -0,0 +1,32 @@
+        self.schema_path = schema_path
+        self.isvalid()
+        
+    def isvalid(self):
```
Look at Maria PR. I commented on this

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,25 @@
+        command = CommandFactory.create_commands(args)
+        result = command.execute()
+    except Exception as e:
+       output_object = Output(e,result)
```
It doesn’t have any issue, but to avoid any mistake, I recommend pass the parameter keys as well
result=result, exception = e

mohamedabdelsalam commented on output/output.py:
```diff
@@ -0,0 +1,15 @@
+        else :
+            self.message = "the command has been executed successfully!"
+            self.status = Status.printName(Status.SUCCESS)
+        self.json_obj = {"status": self.status, "message": self.message, "result": result}
```
Why do u need self.json_obj ??

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,25 @@
+        output_object = Output(result)
+
+    output_josn = output_object.output_json()
+    print(json.dumps(output_josn, indent = 2)) 
```
You can print (json.dump(obj.__dict())
Search for how to serilize object in python. You will find something like that

josephShokry committed:
cleaning the main & adding the super of exceptions

cleaning the main by removing my implementation of the dict and you the built in, adding the super to the exceptions and removing the message attribute

josephShokry committed:
Update main.py

josephShokry committed:
Update create_command.py

josephShokry committed:
remove the icommand

mohamedabdelsalam commented on Commands_Functions/create_command.py:
```diff
@@ -0,0 +1,30 @@
+
+class CreateCommand():
+    def __init__(self, schema_path):
+        self.schema_data = self.isvalid(schema_path = schema_path)
```
Why you didn’t make isValid static ?
IsSomething means it returns a Boolean. I recommend rename it

mohamedabdelsalam commented on Commands_Functions/delete_command.py:
```diff
@@ -0,0 +1,10 @@
+
+class DeleteCommand():
+
+    def isvalid(self):
```
Remove isValid from another classes including this

mohamedabdelsalam commented on main.py:
```diff
@@ -0,0 +1,22 @@
+from Commands_Functions.command_factory import CommandFactory
+
+if __name__=='__main__':
+    result = None
```
Why do you have a result = None here and u are using it only inside try ?

mohamedabdelsalam commented on output/status.py:
```diff
@@ -0,0 +1,12 @@
+    MissingInput = 3
+    FileNotFound = 4
+    WrongInput = 5
+
```
Why do you need printName. You can use .name directly.

also print means printing not returning a string.

hussainmansour committed:
replay and solve comments

mohamedabdelsalam reviewed (commented):
remove __pycache folder

mohamedabdelsalam commented on Commands_Functions/create_command.py:
```diff
@@ -0,0 +1,32 @@
+        self.schema_data = CreateCommand.__validate(schema_path)
+
+    @staticmethod  
+    def __validate(schema_path):
```
Validate expects to return true/false or throws an exception. I would rename this method by `__validate_and_get_schema`

mohamedabdelsalam commented on README.md:
```diff
@@ -1,2 +1,2 @@
 # SimpleFSDB
-simple file system database 
+simple file system database 
```
remove all .pyc. I recommend add `.gitignore` file and add `*.pyc`

Look at Mariam PR. she did the same

josephShokry committed:
adding gitignore file

ignore pyc files

josephShokry committed:
Delete __pycache__ directory

josephShokry committed:
Delete output/__pycache__ directory

josephShokry committed:
Delete Commands_Functions/__pycache__ directory

Commit:
testing

Commit:
testing

josephShokry committed:
folders tests

josephShokry committed:
Merge branch 'test-case' into Handling-Errors
